python/mxnet/autograd.py

```diff
@@ -112,6 +137,34 @@ def pause(is_train=False):
     return RecordingStateScope(False, is_train)
 
 
+def override_train():
```

Comment:
Thanks @szha for working hard to resolve these issues.
I really like that you've added setters / getters for training and recording and am relieved to see them decoupled. 

So one question is - 
Is predict =True equivalent to training=False? 
If so then we would only need to set training and I'd agree with Eric that we could hold off on adding predict() right now for fear of introducing redundancy. Still we have a tough situation for a modeler who wants at will to be able to run forward passes with dropout on vs off. It's hard to see how they can turn dropout on/off without messing with batchnorm.

For me, the main idea for having a predict(): scope would be *if the default were training=True*. I tend to think that tis should be the case. 

**Argument: If you just instantiate a Dropout layer and call forward on it, you'd expect the Dropout to be applied**

At the same time I'm cognizant of Eric's belief that this breaks from the previous way of doing things and might harm backwards compatibility in some sense.